src/types/query.d.ts

```diff
@@ -88,29 +251,53 @@ export module query {
   ): Expr
   export function Remove(ref: ExprArg, ts: ExprArg, action: ExprArg): Expr
   export function CreateClass(params: ExprArg): Expr
-  export function CreateCollection(params: ExprArg): Expr
+
+  export function CreateCollection<T extends object, Meta extends object = any>(
+    params: ExprVal<CreateCollectionParams<Meta>>
+  ): Expr<Collection<T, Meta>>
+
+  // TODO
   export function CreateDatabase(params: ExprArg): Expr
-  export function CreateIndex(params: ExprArg): Expr
+
+  export function CreateIndex<T extends object, Meta extends object = any>(
+    params: ExprVal<CreateIndexParams<Meta>>
+  ): Expr<Index<T, Meta>>
+
+  // TODO
   export function CreateKey(params: ExprArg): Expr
-  export function CreateFunction(params: ExprArg): Expr
+
+  export function CreateFunction<Return, Meta extends object = any>(
+    params: ExprVal<FunctionParams<Meta>>
+  ): Expr<Function<Return>>
+
+  // TODO
   export function CreateRole(params: ExprArg): Expr
 
+  // TODO
   export function Singleton(ref: ExprArg): Expr
   export function Events(ref_set: ExprArg): Expr
-  export function Match(index: ExprArg, ...terms: ExprArg[]): Expr
+
+  export function Match<T extends object>(
+    index: Expr.IndexRef<T>,
+    ...terms: any[]
+  ): Expr.SetRef<Document<T>>
```

Comment:
this is wrong as well, it should return `SetRef<T>`, because when materialized that set won't return documents, but it will depends on the index values

Reply:
yeah, I've assumed `IndexRef` to contain documents only for now, because I'm not using custom values yet. 

Contribution welcome!